### ISSUE TITLE
Performance of (alpha) UMLS Linker on MedMentions ?

### ISSUE THREAD

**Issue by danlou** [allenai/scispacy#134]
Hi,

Just wondering if you have any results regarding the performance of the alpha UMLS linker on MedMentions, using the mention span detector you trained for en_core_sci_*.

Also, I see an F1 of 69.26 on https://allenai.github.io/scispacy/, this is the span detection performance of the latest en_core_sci_md on the test set for full data right (not st21pv) ? Is data from st21pv ever used?

The MedMentions paper only reports results for TaggerOne with a very beefy setup (0.9TB RAM!), it'd be really useful to have another baseline using scispacy, even in this alpha stage.

Thanks,
Dan

**Comment by DeNeutoy** [comment]
Hi Dan,

Thanks - are you specifically interested in the performance when combined with the mention span detector (i.e using predicted mention spans and not gold mention spans)?

Yes, the 69.26 F1 is the span detection F1 for the latest en_core_sci_md/sm (I can't actually remember which one, but they are essentially the same). However, this does not include performance when considering the semantic types predicted by the model, because we don't actually use those at all yet in the linker.

You are correct that the st21pv is not used at all.

Wow I did not know that TaggerOne requires 1TB RAM 😭 😭 😭 


I've added some statistics below based on our development version of the linker, which was not integrated into the library. As such there may be small differences in numbers:

#### Document level scores

This evaluation is at the _document level_, based on the "bag of entities" in an abstract. I.e we are measuring the recall of gold concepts in the aggregated list of all mention candidates in the document. This is meaningful because quite often, we actually care about document level metrics (search, displaying in a UI) in practice. Additionally, it helps with the abbreviation problem, because quite often a mention will be defined in full and then used as an abbreviation.

*Base Configuration*
- No abbreviation substitution
- Mentions generated based on `en_core_web_md`

K | Base | +Noun Chunks | Gold Mentions
-- | -- | -- | --
1 | 57.34 | 60.41 | 72.35
2 | 62.6 | 65.96 | 78.52
10 | 71.73 | 75.83 | 86.33
40 | 78.22 | 82.3 | 91.13
60 | 79.85 | 83.91 | 92.31
80 | 80.65 | 84.66 | 92.79
100 | 81.17 | 85.19 | 93.13

#### Soft Matching Mention level scores
- No abbreviation substitution
- Mentions generated based on `en_core_web_md`
- predicted mentions are compared based on a "soft overlap", meaning that we take it to be a match with any gold mention with which it overlaps at all.

K | Recall@K
-- | --
1 | 49.36
2 | 54.27
10 | 62.64
40 | 68.62
60 | 70.09
80 | 70.91
100 | 71.49


Finally, it's also helpful to measure _normalised_ recall@1, where we consider only results for which the gold mention is present in the generated candidates for a given K. You can get an approximation of this just by dividing K=1/K=N (this is not quite accurate as it's possible that when we throw away entities which we can't generate candidates for, the set that is left is easier overall, so this is a slight underestimate probably). E.g normalised soft matching with predicted mentions and K=80 == 69.6% Accuracy.

Hopefully those numbers are useful for understanding the performance!

There are many pitfalls when evaluating EL systems, it's really pretty annoying. P,R and F1 really don't give you enough of an idea of how the system is performing as a whole. Let me know if you need to know anything specific about it!

Mark

**Comment by danlou** [comment]
Hi Mark,

Thanks for such a detailed response, it's definitely helpful! Yes, I'm interested in mention span detection and entity linking (concept recognition), under similar conditions as the MedMentions paper (though more interested on the full dataset too).

This evaluation is under slightly different conditions from the MedMentions paper (they focused on st21pv, 'hard' matches) but these results seem to suggest that your lightweight approach (even in alpha) can be competitive with the paper's very impractical baseline (maybe even better for document-level).

I'm also realizing that evaluating EL, particularly for MedMentions, won't be as straightforward as it should be... P/R/F1 are a bit harsh indeed, specially considering that span detection in this dataset seems quite challenging. Even with gold spans I expect that recall@1 should be very hard given the number of closely related concepts (incl. many without annotations in the training set!).

In any case, MedMentions seems like a great dataset otherwise, it's only really lacking detailed evaluation results. It's still very recent though.

I have a couple of questions on this, mostly to make sure I'm understanding everything:
 - When you say that 'we don't actually use [semantic types] at all yet in the linker', are you referring to the concept annotations on MedMentions ? The linker does match predicted spans to UMLS concepts (and thus sematic types), the difference here is that you're only considering the aliases from the UMLS KB, right?

- I don't think I understand the normalised recall@1 you describe. Doesn't evaluating under a scenario where you throw away spans that didn't include the gold mention (concept) lead to overestimating performance instead of underestimating? I also didn't understand how the approximation through K works, and where the accuracy comes from.

Thanks once again, this is great!
Dan

**Comment by DeNeutoy** [comment]
No problem!

In answer to your questions:

1. Right, the concept annotations are not used because we only are looking at string similarity to aliases from the kb.

2. Yes this is a bit complicated. It basically comes down to the following:

There are 3 things we want to evaluate when doing end to end linking: 1) mention detection, 2) candidate generation and 3) choosing the right candidate. It's clear how to ablate the mention detection - just use gold mentions. To do the same thing with respect to the candidate selection part, you can say "for the mentions for which the top K contains the gold concept (i.e we did the generation step successfully), what is my accuracy with respect to this subset?". This separates out performance between the generation step and the selection step. The reason I then took the recall@1/recall@k is that the linker is currently only based on string similarity, so to get a single output you just take the one with the highest score. The reason that this is not quite accurate is that it is possible that the datapoints that you remove for a given K are "hard" in some sense, and the accuracy over this subset for top 1 is actually higher than over the whole dataset (this is likely to be the case), which is why this number is a lower bound of the normalized accuracy. 

I am currently working on some integration stuff with `allennlp` (see https://github.com/allenai/allennlp/pull/3040) to make it easier to hook in more complicated models, and I also want to try the entity linker from spacy too.


I've just re-read the medmentions paper and it seems like they added much more detail about the st21pv subset - previously when I'd read it I thought it was much more specific to IR. I think possibly we should switch over to this subset as it sounds more likely to reflect what people actually want out of an entity linker.

It seems like some useful evaluations would be the following:

- P/R/F1 on the st21pv subset with gold mentions
- mention detection P/R/F1 on the st21pv subset

It would also be helpful to know the coverage of our kb on this subset as well as what % of annotated concepts in this subset have definitions. I will try to work on these things over the next couple of weeks.

**Comment by DeNeutoy** [comment]
Ok, i'm training a mention detector now. It should be done soon, that will give a good first indication about whether removing the other types makes everything easier. The other stuff (with the kb) might take a bit longer, but i'll update this thread as I go.

**Comment by danlou** [comment]
Hi Mark,

Thanks for keeping on providing more details.
We're going to be working on this over the next couple of weeks (with some breaks though), so I think I can help with those evaluations (we should need them too).

I've started experimenting with scibert, converting MedMentions to conll format and training a mention detector with it's default NER model, and I'm seeing a significant improvement on F1 ("test_f1-measure-overall": 0.7646). Still need to confirm a few things though (e.g. https://github.com/allenai/scibert/issues/50).

I'll let you know more as I make progress too. Still not sure if I'm more interested in the full data or st21pv, but am interested to know your results on st21pv in any case.

Thanks

**Comment by kingsaint** [comment]
> Hi Dan,
> 
> Thanks - are you specifically interested in the performance when combined with the mention span detector (i.e using predicted mention spans and not gold mention spans)?
> 
> Yes, the 69.26 F1 is the span detection F1 for the latest en_core_sci_md/sm (I can't actually remember which one, but they are essentially the same). However, this does not include performance when considering the semantic types predicted by the model, because we don't actually use those at all yet in the linker.
> 
> You are correct that the st21pv is not used at all.
> 
> Wow I did not know that TaggerOne requires 1TB RAM 😭 😭 😭
> 
> I've added some statistics below based on our development version of the linker, which was not integrated into the library. As such there may be small differences in numbers:
> 
> #### Document level scores
> This evaluation is at the _document level_, based on the "bag of entities" in an abstract. I.e we are measuring the recall of gold concepts in the aggregated list of all mention candidates in the document. This is meaningful because quite often, we actually care about document level metrics (search, displaying in a UI) in practice. Additionally, it helps with the abbreviation problem, because quite often a mention will be defined in full and then used as an abbreviation.
> 
> _Base Configuration_
> 
> * No abbreviation substitution
> * Mentions generated based on `en_core_web_md`
> 
> K	Base	+Noun Chunks	Gold Mentions
> 1	57.34	60.41	72.35
> 2	62.6	65.96	78.52
> 10	71.73	75.83	86.33
> 40	78.22	82.3	91.13
> 60	79.85	83.91	92.31
> 80	80.65	84.66	92.79
> 100	81.17	85.19	93.13
> #### Soft Matching Mention level scores
> * No abbreviation substitution
> * Mentions generated based on `en_core_web_md`
> * predicted mentions are compared based on a "soft overlap", meaning that we take it to be a match with any gold mention with which it overlaps at all.
> 
> K	Recall@K
> 1	49.36
> 2	54.27
> 10	62.64
> 40	68.62
> 60	70.09
> 80	70.91
> 100	71.49
> Finally, it's also helpful to measure _normalised_ recall@1, where we consider only results for which the gold mention is present in the generated candidates for a given K. You can get an approximation of this just by dividing K=1/K=N (this is not quite accurate as it's possible that when we throw away entities which we can't generate candidates for, the set that is left is easier overall, so this is a slight underestimate probably). E.g normalised soft matching with predicted mentions and K=80 == 69.6% Accuracy.
> 
> Hopefully those numbers are useful for understanding the performance!
> 
> There are many pitfalls when evaluating EL systems, it's really pretty annoying. P,R and F1 really don't give you enough of an idea of how the system is performing as a whole. Let me know if you need to know anything specific about it!
> 
> Mark

Suppose I have the gold mentions available from Oracle. How can I provide the gold mentions to the entity linking model during inference? In other words, is there a way to avoid mention detection?

**Comment by danlou** [comment]
FYI, I've since explored this further on a short paper at ECIR. The (preliminary) code is available [here](https://github.com/danlou/MedLinker).

The MedLinker codebase is quite modular, and allows for providing gold mentions. It was used to produce results with gold mentions when writing the paper. The [predict()](https://github.com/danlou/MedLinker/blob/master/medlinker.py#L112) method in the MedLinker class accepts gold spans as a parameter.

There's little to no documentation/testing for that project though, and I haven't worked on that project for some time...

It's also possible that there's this functionality in scispacy, in which case you're probably better off relying on that.

**Comment by kingsaint** [comment]
> FYI, I've since explored this further on a short paper at ECIR. The (preliminary) code is available [here](https://github.com/danlou/MedLinker).
> 
> The MedLinker codebase is quite modular, and allows for providing gold mentions. It was used to produce results with gold mentions when writing the paper. The [predict()](https://github.com/danlou/MedLinker/blob/master/medlinker.py#L112) method in the MedLinker class accepts gold spans as a parameter.
> 
> There's little to no documentation/testing for that project though, and I haven't worked on that project for some time...
> 
> It's also possible that there's this functionality in scispacy, in which case you're probably better off relying on that.

Thanks, Dan! This is super helpful.
